### ISSUE TITLE
Feature request: zfs rebalance command/property

### ISSUE THREAD

**Issue by tonymmm1** [openzfs/zfs#12267]
### Describe the feature would like to see added to OpenZFS

If a zpool gets expanded with additional vdevs it would be nice if there was either a property flag or a command to rebalance the vdevs. Being able to specify vdevs would be a bonus but as other zpool commands do not target vdevs specifically its probably not necessary. 

``zpool rebalance tank vdev1 vdev2`` 

or

``zpool set -O feature@pool_rebalance=enabled``

### How will this feature improve OpenZFS?

Allows for more easy expansion of zpools without having to copy and rebuild manually. 

### Additional context
Personally I had a pool of 3 - 8TB disks that I had in a mirror vdev that I broke into a striped mirror of 4 - 8TB and now its unbalanced on half by about 1.5TB. I have rebuilt previous expanded zpools manually as stated above but I believe it should be added as a useful feature for systems admins. 

Hopefully this proposal gets acknowledged if not I understand that re-balancing the metadata and recopying the files would add a lot more complexity and might not be as trivial as it sounds.

**Comment by bghira** [comment]
i too would love to change my pants while running at full speed.

**Comment by kernelOfTruth** [comment]
when added, let's just hope it won't be a pre-requisite to have ZFS running (similar to Btrfs)

e.g. when just rsync-ing data to a Btrfs volume after several hundred GiB or a few TiB I would have to juggle and balance the data, cause it ran into ENOSPC -_-

**Comment by tonymmm1** [comment]
> when added, let's just hope it won't be a pre-requisite to have ZFS running (similar to Btrfs)
> 
> e.g. when just rsync-ing data to a Btrfs volume after several hundred GiB or a few TiB I would have to juggle and balance the data, cause it ran into ENOSPC (and that's with a single device) -_-
> 
> so far I haven't encountered anything similar on ZFS _knock on wood_

I was more thinking along the lines of it being run manually and afterwards all new data would be distributed among vdevs as normal.

**Comment by kernelOfTruth** [comment]
> > when added, let's just hope it won't be a pre-requisite to have ZFS running (similar to Btrfs)
> > e.g. when just rsync-ing data to a Btrfs volume after several hundred GiB or a few TiB I would have to juggle and balance the data, cause it ran into ENOSPC (and that's with a single device) -_-
> > so far I haven't encountered anything similar on ZFS _knock on wood_
> 
> I was more thinking along the lines of it being run manually and afterwards all new data would be distributed among vdevs as normal.

yep, gotcha - wasn't meant to be demeaning - it just made me think how this potentially isn't that easy

but it could use a similar approach like resilvering / mirroring, raid setups, I assume

on that note: it sounds quite useful :+1:

**Comment by IvanVolosyuk** [comment]
I believe the issue is no go without block pointer rewrite. ZFS never moves written data, except for https://github.com/openzfs/zfs/pull/12225 which also doesn't change block pointers. So far you can use 'zfs send | zfs receive' for a dataset to balance it offline if you have enough free space. BTRFS has enough indirection to make this happen which makes it more complex file system, harder to develop IMHO.

**Comment by NicholasRush** [comment]
ZFS rebalancing  would be very nice, but it should work like the way, like a storage optimization service. That means if you change dataset properties like dedup or compression and you add some new vdevs the data should be optimized in background automatically if the system have less load. This could be triggered automatically or started and aborted by a command like "zfs optimize" and paused by "zfs optimize abort" and should be triggered to work in foreground mode for datasets or volumes and with "zpool optimize...." at pool level in background mode.

**Comment by joshenders** [comment]
Looks like this feature got some attention at the 2023 dev summit in San Francisco. @pcd1193182, I've searched but haven't come across any info, do you have anything you can share? Seems well timed with the vdev expansion merge happening soon.
